Charts

Kody91 opened the pull request.

Commit:
Still needs a bit more work, but I think this line chart implementation
should work for our needs.

TODO: need to externalize the colours of the Line.

Commit:
extracted line colours to avoid users from having to use the QT
namespace.

Started removing earlier portions from the line to ensure that the list \
does not get too large and effect performance

Commit:
Properly annotated problem areas. Need to better understand QT api, or
add new variables into object to find proper min/max without iterating
through a bunch of lists.

Commit:
Modified line chart into an area chart, as per client requirements

Commit:
added pyqtchart bindings into requirements

Commit:
Added some unit tests to line chart module

Commit:
Made changes to code to fix issues made apparent by unit testing

Commit:
Added pyqtchart to setup.py

Commit:
Replaced Qtcharts with the correct QtChart name from pip

Commit:
Added missing comma inside the setup.py file

Commit:
Added Display and XAuthority to passenv of Tox.ini, to fix linux specific testing issue.

Commit:
Added pytest-xvfb into the testenv requirements to use a virtual X
server buffer to conduct tests

Commit:
Did proper init

Commit:
Renamed to chart.py

Commit:
Chart rename

davidjoseph9 commented on hmpy/views/chart.py:
```diff
@@ -0,0 +1,115 @@
+    def _get_y_limits(self):
+        list_of_y_points = []
+        point_list = self._line_actual.pointsVector()
+        for i in point_list:
```
`    """
        :returns tuple, first element being the minimum Y value in the series, second element is the max Y value in the
        series
    """
    def _get_y_limits(self):
        list_of_y_points = []
        return _min_max_y(self._line_actual.pointsVector())
    """
        :returns tuple, first element being the minimum X value in the series, second element is the max X value in the
        series
    """
    def _get_x_limits(self):
        list_of_x_points = []
        return _min_max_x(self._line_actual.pointsVector())


def _min_max_x(data):
    if len(data) > 0:
        lo = hi = data[0].x()

        for next in data:
            if next.x() < lo:
                lo = next.x()
            if next.x() > hi:
                hi = next.x()
    else:
        raise ValueError('min_max() data is an empty sequence')
    return lo, hi


def _min_max_y(data):
    if len(data) > 0:
        lo = hi = data[0].y()

        for next in data:
            if next.y() < lo:
                lo = next.y()
            if next.y() > hi:
                hi = next.y()
    else:
        raise ValueError('min_max() data is an empty sequence')
    return lo, hi
`

Commit:
Changes to charts

Commit:
Removed references to Line from init file

Commit:
Refactored tests to now reflect new structure of chart class.

Commit:
Refactored charts to now keep the Line item internal to the chart class.

Small refactor to limits methods. Changed some names, and removed the Line getter, as it's no longer needed.

Commit:
refactored limits function to now guarantee an O(n) time. Should improve
performance a little bit.

davidjoseph9 reviewed (dismissed):
Good Job!

jseed committed:
Add scale option to View constructor

Commit:
Added in width to the init params of chart. Renamed width to
line_width which seems a little more intuitive

Commit:
Fixed mixed axises
